### DISCUSSION TITLE
crop clarification

### DISCUSSION THREAD

**Discussion by jriker1** [quietvoid/dovi_tool#173] [General]
With the crop function, I am looking to take 4k HDR10/Dolby Vision content and re-encode it and crop off the black bars.  Below was my workflow and thinking it's now inaccurate after some conversations elsewhere: 

- demux original video track to HEVC (let's say it's 3840 x 2160 with 280 of black bars on top and bottom.)
- Run: dovi_tool extract-rpu <video>.h265
- Run: ffplay -i myfile -vf cropdetect to find the crop marks in this case 280 top and bottom.
- Run to convert: 
NVEncC64.exe --avhw --input <input_filename>.mkv --output <output_filename>.hevc --codec h265 --tier high --level 5.1 --output-depth 10 --cqp 19 --preset quality --vbv-bufsize 15000 --lookahead 32 --bframes 4 --aq-temporal --aud --repeat-headers --chromaloc 2 --colorprim bt2020 --transfer smpte2084 --colormatrix bt2020nc --colorrange limited --max-cll 413,65 --master-display copy --video-metadata copy --crop 0,280,0,280 --gop-len 0 --ref 3 --profile main10 --bref-mode middle --mv-precision Q-pel

- Run: dovi_tool inject-rpu -i video.hevc --rpu-in RPU.bin -o injected_output.hevc

I assume the above is incorrect however is it as simple as adding the -c option on the last inject step above?

**Comment by quietvoid** [comment]
> I assume the above is incorrect however is it as simple as adding the -c option on the last inject step above?

No. It should either be added to the first `extract-rpu` step or you can also edit the RPU afterwards.
Most global options have no effect when injecting.

Assuming you're reencoding a UHD BD, you also need to convert using mode 2.

**Comment by manuelrn** [comment]
Simply extract the RPU with this command, as @quietvoid says, and you don't worry about anything else.

`dovi_tool -m 2 --crop extract-rpu`

Although if the source has a variable size of black bars, for example movies that include imax scenes, then it is not so simple in case you plan to resize the video resolution, if you are going to keep it at 4K then it is as easy as not cropping the source and not cropping the RPU.

**Comment by jriker1** [comment]
thanks for the replies.  How does the -m 2 and --crop work in this case?  So the black bars still exist in the source running dovi_tool against, and the black bar removal size could vary per video.  Does dovi_tool identify the black bars like 

ffplay -i myfile -vf cropdetect 

calculates?

**Reply by quietvoid** [reply]
No, it only cares about whether there are still black bars or not.
Specifying `--crop` tells it that there aren't any in the final video.

If the bars vary in size, the accurate way is to edit the RPU properly.  
But this is more complicated.

**Reply by jriker1** [reply]
Thanks.  I probably or don't understand how dolby vision data is stored so was thinking if one video has 280 pixels of black bars on the top and bottom and another has 180 on top and bottom, aren't those relevant when using the crop command in dovi_tool?  I was thinking of Dolby Vision like it mapped each frame colors per pixel or something and to know how much was going to be cut off the top and bottom it would have to know what that number is.

Do notice the bin file is smaller with  -m 2 --crop, 38 megs vs 45 megs.

**Comment by quietvoid** [comment]
> aren't those relevant when using the crop command in dovi_tool?

Not when using crop.

The metadata only stores the letterbox bars size (with values for left, right, top, bottom).
When using `--crop`, these are all set to zero to specify that there are no bars in the video.

If the file still has letterbox bars of different size, then the metadata should be adjusted accordingly.
